app/src/main/java/io/appium/settings/RecorderService.java

```diff
@@ -0,0 +1,138 @@
+    }
+
+    @Nullable
+    @Override
+    public IBinder onBind(final Intent intent) {
+        return null;
+    }
+
+    @SuppressLint("NewApi")
+    @Override
+    public int onStartCommand(final Intent intent, final int flags, final int startId) {
+        Log.v(TAG, "onStartCommand:intent=" + intent);
+
+        int result = START_STICKY;
+        final String action = intent != null ? intent.getAction() : null;
+        Log.v(TAG, "onStartCommand:intent=" + "action string: " + action);
+        if (ACTION_RECORDING_START.equals(action)) {
+            showNotification(TAG); // TODO is this really necessary
+
+            MediaProjectionManager mMediaProjectionManager =
+                    (MediaProjectionManager) getSystemService(Context.MEDIA_PROJECTION_SERVICE);
+
+            startScreenRecord(mMediaProjectionManager, intent);
+        } else if (ACTION_RECORDING_STOP.equals(action)) {
+            Log.v(TAG, "onStartCommand:intent=" + "stop");
+            stopScreenRecord();
+            result = START_NOT_STICKY;
+        }
+
+        return result;
+    }
+
+    /**
+     * start screen recording
+     */
+    @SuppressLint("NewApi")
+    private void startScreenRecord(MediaProjectionManager mediaProjectionManager,
+                                   final Intent intent) {
+        if (recorderThread == null) {
+            Log.v(TAG, "recorder thread null");
+
+            final int resultCode = intent.getIntExtra(ACTION_RECORDING_RESULT_CODE, 0);
+            // get MediaProjection
+            final MediaProjection projection = mediaProjectionManager.getMediaProjection(resultCode,
+                    intent);
+            if (projection != null) {
+                Log.v(TAG, "projection different from null");
+                final DisplayMetrics metrics = getResources().getDisplayMetrics();
+                int width = metrics.widthPixels;
+                int height = metrics.heightPixels;
+                if (width > height) {
+                    // Horizontal
+                    final float scale_x = width / 1920f;
+                    final float scale_y = height / 1080f;
+                    final float scale = Math.max(scale_x,  scale_y);
+                    width = (int)(width / scale);
+                    height = (int)(height / scale);
+                } else {
+                    // Vertical
+                    final float scale_x = width / 1080f;
+                    final float scale_y = height / 1920f;
+                    final float scale = Math.max(scale_x,  scale_y);
+                    width = (int)(width / scale);
+                    height = (int)(height / scale);
+                }
+                Log.v(TAG, String.format("startRecording:(%d,%d)(%d,%d)",
+                        metrics.widthPixels, metrics.heightPixels, width, height));
+                String outputFilePath = intent.getStringExtra(ACTION_RECORDING_FILENAME);
+                if (outputFilePath != null) {
```

Comment:
shall we check if recording is already running?
shall we log an error if no output file path is provided?

Reply:
we are checking status of recording with nullability currently but i can change to helper method(isRunning) 👍  and `outputFilePath` null check in here is just an extra check to solidify code, it must not be null in all conditions because we are supplying default file name in here https://github.com/sirmordred/io.appium.settings/blob/master/app/src/main/java/io/appium/settings/Settings.java#L107 but again i can add extra check 👍